Fix backlight and add colour temperature to certain Kobo models

GeekGoneOld opened the pull request:
<!--

Thank you for your interest in contributing to XCSoar! Please read the
following information to make it easier for us to review your changes.

We appreciate if you make sure to:

  - Document the changes (in the NEWS.txt file, and the manual when relevant)
  - Enable maintainer edits[1] (in case we need to help with the PR)
  - Check your commits and their messages (so they're all tidy and coherent)

[1] https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->


Brief summary of the changes
----------------------------

<!--
Please note that the commit messages is where detailed descriptions of the
changes should be made - this section is just for a brief summary/overview.
-->
Added CLARA HD backlight, CLARAs HD and 2E and LIBRA 2 colour temp

Related issues and discussions
------------------------------

<!--
Please link any relevant issues or forum posts here, for reference.

If this PR resolves an existing issue, please write "Closes #1234" so that
the issue is closed automatically when this PR is merged.
-->

lordfolken reviewed (changes requested):
While your change here is minimally invasive, it also adds to the unecessary complexity.

Can you 
* rewrite the GetInfo function as a case statement matching the model similar as in system.cpp?
* eliminate the clones by creating a function passing the path

The commit messages would be good if you can add kobo: to the beginning.

ubx commented:
It does not work on my Libra 2:
```
~ # ls /sys/class/backlight/lm3630a_led/color
ls: /sys/class/backlight/lm3630a_led/color: No such file or directory
```
but it would (I have tried it):
```
~ # ls /sys/class/leds/aw99703-bl_FL1/color 
/sys/class/leds/aw99703-bl_FL1/color
```
there are probably several HW revisions.
IMHO you should check if the path exists (maybe there are different possibilities)  before display the color menu.

Here my Libra 2 config:
``` 
~ # /bin/ntx_hwconfig -s /dev/mmcblk0
HW CONFIG v3.5 @1024 secno, size=74 bytes : 
[0] PCB='E70K10'
[1] KeyPad='PGUP+PGDN'
[2] AudioCodec='No'
[3] AudioAmp='No'
[4] Wifi='RTL8723DS'
[5] BT='RTL8723DS'
[6] Mobile='No'
[7] TouchCtrl='ekth6'
[8] TouchType='C-Type'
[9] DisplayCtrl='MX6SLL+SY7636'
[10] DisplayPanel='7" Right EPD'
[11] RSensor='KX122'
[12] MicroP='No'
[13] Customer='9'
[14] Battery='PR-284983N-1500mA'
[15] Led='W'
[16] RamSize='512MB'
[17] IFlash='Micro SD'
[18] ExternalMem='No'
[19] RootFsType='Ext4'
[20] SysPartType='TYPE3'
[21] ProgressXHiByte=0x01
[22] ProgressXLoByte=0x68
[23] ProgressYHiByte=0x02
[24] ProgressYLoByte=0xE4
[25] ProgressCnts=0x00
[26] ContentType=0x00
[27] CPU='mx6sll'
[28] UIStyle='Customer UI'
[29] RAMType='LPDDR2'
[30] UIConfig='Normal'
[31] DisplayResolution='1680x1264'
[32] FrontLight='TABLE0+'
[33] CPUFreq='NC'
[34] HallSensor='TLE4913'
[35] DisplayBusWidth='16Bits'
[36] FrontLight_Flags=BootON:OFF,TABLE1X:OFF,EN_INV:ON,(null):OFF,(null):OFF,(null):OFF,(null):OFF,(null):OFF
[37] PCB_Flags=NO_KeyMatrix:ON,FPC_Touch:OFF,LOGO_LED:OFF,RD_MODE:OFF,EPD_LV:OFF,EPDVDD_STDALONE:OFF,HP_DETECT:OFF,ISD_1V8:OFF
[38] FrontLight_LEDrv='LM3630'
[39] VCOM_10mV_HiByte=0xFF
[40] VCOM_10mV_LoByte=0x6A
[41] PCB_REV=0x00
[42] PCB_LVL='E'
[43] HOME_LED_PWM='No'
[44] PMIC='BD71828'
[45] FL_PWM='AW99703x2'
[46] RTC='BD71828'
[47] BootOpt=ESD:OFF
[48] Touch2Ctrl='No'
[49] Touch2Type='No'
[50] GPS='No'
[51] FM='No'
[52] RSensor2='No'
[53] LightSensor='No'
[54] TPFWIDByte0=0x00
[55] TPFWIDByte1=0x00
[56] TPFWIDByte2=0x00
[57] TPFWIDByte3=0x00
[58] TPFWIDByte4=0x00
[59] TPFWIDByte5=0x00
[60] TPFWIDByte6=0x00
[61] TPFWIDByte7=0x00
[62] GPU='NC'
[63] PCB_Flags2=eMMC@SD1:ON,WiFi@SD2:OFF,eMMC@SD2:OFF,USB_TYPEC:ON,Cilix:OFF,(null):OFF,(null):OFF,(null):OFF
[64] EPD_Flags=NO_VDD_CTRL:OFF,LP_PANEL:OFF,ACEP:OFF,CFA:OFF,XON:OFF,(null):OFF,(null):OFF,(null):OFF
[65] LAN='No'
[66] MobileIF='NC'
[67] PIR='No'
[68] PanelLaminationSrc='NC'
[69] FL_CHNS=0x04
[70] TTS='No'
[71] Test_Flags=0:OFF,1:OFF,2:OFF,3:OFF,4:OFF,5:OFF,6:OFF,7:OFF
[72] EPD_CFA_REV='Gen1'
[73] PanelModel='NC'
```

MaxKellermann commented on src/Kobo/System.cpp:
```diff
@@ -424,3 +424,65 @@ KoboSetBacklightBrightness([[maybe_unused]] int percent)
+{
+#ifdef KOBO
+  switch (DetectKoboModel()) {
+    case KoboModel::LIBRA2:
+    case KoboModel::CLARA_2E:
+    case KoboModel::CLARA_HD:
```
Bad indent

GeekGoneOld replied:
Fixed

MaxKellermann commented on src/Kobo/System.hpp:
```diff
@@ -71,3 +71,12 @@ KoboGetBacklightBrightness();
+bool
+KoboCanChangeBacklightColour();
+
+int
+KoboGetBacklightColour();
```
What does the return value mean? Why is it signed? Can it throw exceptions?

GeekGoneOld replied:
The return value is an integer from 0 to 10 where 0 is lowest colour temperature (yellowish) and 10 is highest (white/blueish).  The actual number is not any known scale (like degrees K or percent).

The return value is only signed because it gets its value from atoi() which returns int.  Do you have a preferred designation (uint8_t?).  If you wish me to change it, would you like me to change the Brightness code (similar but 0-100)  to your preferred type as well?

I don't see how it could throw an exception.  It reads a string from the driver (which is noexcept) and calls atoi() which returns 0 in the case of invalid.

MaxKellermann replied:
The explanation belongs in a doxygen-style code comment.

GeekGoneOld replied:
Will gladly do this.  Can you point me to a preferred doxygen sample as I would like to conform to this repo's standard?  Do you want the entire function def documented or just the colour return value (get) / call parameter (set)?

Also, would you like me to document both functions (set and get) as they both use colour?

GeekGoneOld replied:
I went ahead and did the code comment in the same style as some other (src/Engine/GlideSolvers/GlidePolar.cpp).  I commented both KoboGetBacklightColour and KoboSetBacklightColour and KoboGetBacklightColourFile.  I also added noexcept to all the ...BacklightColour... functions as none of them can cause an exception other than OOM.

MaxKellermann commented on src/Kobo/System.hpp:
```diff
@@ -71,3 +71,12 @@ KoboGetBacklightBrightness();
 void
 KoboSetBacklightBrightness(int percent);
+
+bool
+KoboCanChangeBacklightColour();
```
Can this method throw exceptions?

GeekGoneOld replied:
I don't see how it could throw an exception. It reads a string from the driver (which is noexcept) and calls atoi() which returns 0 in the case of invalid.

GeekGoneOld replied:
I added noexcept

MaxKellermann commented on src/Kobo/SystemDialog.cpp:
```diff
@@ -46,13 +46,17 @@ class SystemWidget final
-    DECREASE_BACKLIGHT_BRIGHTNESS
+    DECREASE_BACKLIGHT_BRIGHTNESS,
+    INCREASE_BACKLIGHT_COLOUR_TEMPERATURE,
+    DECREASE_BACKLIGHT_COLOUR_TEMPERATURE
```
Trailing comma missing

GeekGoneOld replied:
Are you sure?  Line 49 (previously last in list) got a comma added.  Line 51 (new last in list) shouldn't have one.

MaxKellermann replied:
Yes, I'm sure. The comma was missing in commit 79b80960035 but this wasn't noticed in code review.

GeekGoneOld replied:
Added the comma

MaxKellermann commented on src/Kobo/SystemDialog.cpp:
```diff
@@ -224,6 +244,33 @@ SystemWidget::UpdateBacklightButtons(int percent)
+SystemWidget::DecreaseBacklightColourTemperature()
+{
+  int current_colour = KoboGetBacklightColour();
+  KoboSetBacklightColour(current_colour - 2);
```
What if the getter has failed? Why call the setter if it has failed?

GeekGoneOld replied:
I don't see how the getter could fail.  The getter is now tested for presence of the file prior to enabling the buttons.

GeekGoneOld replied:
OK.  I've fixed it to silently return if the getter fails.

MaxKellermann commented on src/Kobo/SystemDialog.cpp:
```diff
@@ -224,6 +244,33 @@ SystemWidget::UpdateBacklightButtons(int percent)
+SystemWidget::UpdateBacklightColourButtons(int colour)
+{
+  if(decrease_backlight_colour_temperature != nullptr) {
+    decrease_backlight_colour_temperature->SetEnabled(colour != 0);
```
This enables the button if the color is negative

GeekGoneOld replied:
Yes.  Good point.  This has been changed to be colour > 0

MaxKellermann replied:
No. You made a mistake: you made the colour a *signed* integer. Why did you choose to make it signed, why did you choose to allow negative numbers?

GeekGoneOld commented on src/Kobo/System.hpp:
```diff
@@ -71,3 +71,12 @@ KoboGetBacklightBrightness();
+int
+KoboGetBacklightColour();
+
+void
+KoboSetBacklightColour(int colour);
```
If we are out of memory, it could throw an exception on std::to_string().  Would you like me to handle that in some specific way?  noexcept?

MaxKellermann replied:
No. OOM is never to be handled. All is lost on OOM anyway.
The fact that you didn't mark this function `noexcept` indicates that this function can throw exceptions, but it's not documented.

GeekGoneOld replied:
changed to noexcept

GeekGoneOld commented:
In reply to @lordfolken (way up near the top of the issue)
>     * rewrite the GetInfo function as a case statement matching the model similar as in system.cpp?
> 
done
>     * eliminate the clones by creating a function passing the path
> 
done (for colour)
> The commit messages would be good if you can add kobo: to the beginning.
>
done

GeekGoneOld commented:
In response to @ubx near the top of this issue (just below @lordfolken ):
> It does not work on my Libra 2:
> 

> There are probably several HW revisions, [see here](https://github.com/NiLuJe/koreader/blob/3841c7b1b18f7a0ec348d1fe3dfdf53a53641c5a/frontend/device/kobo/device.lua).
>
> IMHO you should check if the path exists (maybe there are different possibilities) before display the color menu.

Looked at your link and handled the "silent" model changes much better.  Should work now.

GeekGoneOld commented:
Small change to allow build-ubuntu and build-sanitizer to pass.  Took a couple of trys as I introduced a codacity error that I also fixed.  build-macos still fails as it seems to with every recent PR.

MaxKellermann commented on src/Hardware/Battery.cpp:
```diff
@@ -43,11 +43,13 @@ GetInfo() noexcept
 
-  if (kobo_model == KoboModel::GLO_HD) {
+  switch (DetectKoboModel())
+  {
+  case KoboModel::GLO_HD:
+  case KoboModel::CLARA_HD:
```
Don't do code refactoring in the middle of a patch that is supposed to do something else. This causes a confusing commit diff which is hard to review. You can refactor code, but do it in a separate commit. In this diff, I can't see what really belongs to the backlight feature.

GeekGoneOld replied:
Yes, it grew into confusion.  I have split the PR into 3 well defined commits (the first being refactoring).

MaxKellermann commented on src/Kobo/System.cpp:
```diff
@@ -359,12 +359,14 @@ KoboCanChangeBacklightBrightness()
-      return true;
-    default:
-      return false;
+  case KoboModel::GLO_HD:
+  case KoboModel::LIBRA2:
+  case KoboModel::CLARA_2E:
+  case KoboModel::CLARA_HD:
```
This indent change has nothing to do with backlight, does it?

GeekGoneOld replied:
It does have to do with the backlight but backlight brightness, not backlight colour.  I have done all the refactoring in a separate commit that preceeds the additions so it should be easy to follow.

MaxKellermann commented on src/Kobo/System.cpp:
```diff
@@ -378,20 +380,23 @@ KoboGetBacklightBrightness()
-      }
-      break;
-    default:
-      // nothing to do here...
-      break;
+  case KoboModel::GLO_HD:
+    if (File::ReadString(Path("/sys/class/backlight/mxc_msp430_fl.0/brightness"), line, sizeof(line))) {
+      result = atoi(line);
+    }
+    break;
+
+  case KoboModel::LIBRA2:
+  case KoboModel::CLARA_2E:
+  case KoboModel::CLARA_HD:
+    if (File::ReadString(Path("/sys/class/backlight/mxc_msp430.0/brightness"), line, sizeof(line))) {
+      result = atoi(line);
+    }
+    break;
+
+  default:
+    // nothing to do here...
+    break;
   }
```
More indent changes

MaxKellermann commented on src/Kobo/System.cpp:
```diff
@@ -408,16 +413,124 @@ KoboSetBacklightBrightness([[maybe_unused]] int percent)
-      break;
-    case KoboModel::LIBRA2:
-    case KoboModel::CLARA_2E:
-      File::WriteExisting(Path("/sys/class/backlight/mxc_msp430.0/brightness"), std::to_string(percent).c_str());
-      break;
-    default:
-      // nothing to do here...
-      break;
+  case KoboModel::GLO_HD:
+    File::WriteExisting(Path("/sys/class/backlight/mxc_msp430_fl.0/brightness"), std::to_string(percent).c_str());
+    break;
+
+  case KoboModel::LIBRA2:
+  case KoboModel::CLARA_2E:
+  case KoboModel::CLARA_HD:
+    File::WriteExisting(Path("/sys/class/backlight/mxc_msp430.0/brightness"), std::to_string(percent).c_str());
+    break;
```
Yet more confusing indent changes

GeekGoneOld replied:
Also moved to the refactor commit

MaxKellermann commented on src/Kobo/SystemDialog.cpp:
```diff
@@ -63,7 +67,10 @@ class SystemWidget final
   void IncreaseBacklightBrightness();
   void DecreaseBacklightBrightness();
-  void UpdateBacklightButtons(int percent);
+  void UpdateBacklightBrightnessButtons(int percent);
```
What does the parameter mean? Prior to the rename, it was clear, but it's not any longer.

GeekGoneOld replied:
The backlight has two controls: brightness and colour.  The existing code ignored colour and referred to the backlight brightness as "...Backlight..." in function names  The change to "...BacklightBrightness..." makes it easier to add other backlight controls (like colour) and actually makes the original parameter meaning clearer!

MaxKellermann commented on src/Kobo/SystemDialog.cpp:
```diff
@@ -63,7 +67,10 @@ class SystemWidget final
   void DecreaseBacklightBrightness();
-  void UpdateBacklightButtons(int percent);
+  void UpdateBacklightBrightnessButtons(int percent);
+  void IncreaseBacklightColourTemperature();
+  void DecreaseBacklightColourTemperature();
```
What exceptions do these methods throw? If not at all, add `noexcept`!

GeekGoneOld replied:
I've added noexcept.

MaxKellermann commented:
I could not review any actual code changes this time because this PR had become a confusing mess of refactoring and indentation changes.

GeekGoneOld commented:
Split into 3, organized commits.  Should be far more readable, and reviewable now.

MaxKellermann commented on src/Kobo/System.cpp:
```diff
@@ -426,3 +429,108 @@ KoboSetBacklightBrightness([[maybe_unused]] int percent)
+KoboGetBacklightColourFile() noexcept
+{
+#ifdef KOBO
+  const char * colour_files[3] = {
```
This array should be `const` (or better: `constexpr`)

GeekGoneOld replied:
Yup.  Squeeze all you can into compile!  Changed to constexpr const char * colour_files[3] = {

MaxKellermann replied:
It was initialized at compile time already, but the difference is that it can now be in the `.rodata` section where it will be write-protected.
